thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/alert/content/OnboardingNotificationEmailContentFormatter.java

```diff
@@ -0,0 +1,81 @@
+  public static final String DEFAULT_NULL_STRING_VALUE = "N/A";
+  public static final String ALERT_FILTER_PATTERN_KEY = DetectionJobResource.AUTOTUNE_PATTERN_KEY;
+  public static final String ALERT_CONFIG_NAME = "alertConfigName";
+
+  public OnboardingNotificationEmailContentFormatter() {
+  }
+
+  @Override
+  public void init(Properties properties, ThirdEyeAnomalyConfiguration configuration) {
+    super.init(properties, configuration);
+    this.emailTemplate = properties.getProperty(EMAIL_TEMPLATE, DEFAULT_EMAIL_TEMPLATE);
+  }
+  /**
+   * The actual function that convert anomalies into parameter map
+
+   * @param templateData
+   * @param anomalies
+   */
+  @Override
+  protected void updateTemplateDataByAnomalyResults(Map<String, Object> templateData,
+      Collection<AnomalyResult> anomalies) {
+    AnomalyFunctionDTO anomalyFunctionSpec = null;
+    for (AnomalyResult anomalyResult : anomalies) {
+      if (!(anomalyResult instanceof MergedAnomalyResultDTO)) {
+        throw new IllegalArgumentException("Input anomalies should be instance of MergedAnomalyResultDTO");
+      }
+      if (anomalyFunctionSpec == null) {
+        anomalyFunctionSpec = ((MergedAnomalyResultDTO) anomalyResult).getFunction();
+      } else if (anomalyFunctionSpec.getId() != ((MergedAnomalyResultDTO) anomalyResult).getFunction().getId()) {
+        throw new IllegalArgumentException("Input anomalies should be generated by the same anomaly function");
+      }
+    }
+    templateData.put("functionName", anomalyFunctionSpec.getFunctionName());
+    templateData.put("functionId", anomalyFunctionSpec.getId());
+    templateData.put("metrics", anomalyFunctionSpec.getMetric());
+    templateData.put("filters", returnValueOrDefault(anomalyFunctionSpec.getFilters(), DEFAULT_NULL_STRING_VALUE));
+    templateData.put("dimensionDrillDown", returnValueOrDefault(anomalyFunctionSpec.getExploreDimensions(), DEFAULT_NULL_STRING_VALUE));
+    String alertPattern = DEFAULT_NULL_STRING_VALUE;
+    Map<String, String> alertFilter = anomalyFunctionSpec.getAlertFilter();
+    if (alertFilter != null && alertFilter.containsKey(ALERT_FILTER_PATTERN_KEY)) {
+      alertPattern = alertFilter.get(ALERT_FILTER_PATTERN_KEY);
+    }
+    AlertConfigDTO alertConfig = null;
+    templateData.put("alertPattern", alertPattern);
+    if (templateData.containsKey(ALERT_CONFIG_NAME)) {
+      String configName = (String) templateData.get(ALERT_CONFIG_NAME);
+      alertConfig = DAORegistry.getInstance().getAlertConfigDAO().findWhereNameEquals(configName);
+    }
+    if (alertConfig == null) {
+      alertConfig = new AlertConfigDTO();
+    }
+    templateData.put("application", returnValueOrDefault(alertConfig.getApplication(), DEFAULT_NULL_STRING_VALUE));
+    templateData.put("recipients", returnValueOrDefault(alertConfig.getRecipients(), DEFAULT_NULL_STRING_VALUE));
+  }
+
+  private String returnValueOrDefault(String value, String defaultValue) {
```

Comment:
you can replace this with apache commons MapUtils

Reply:
The problem here is the value is from alertConfig (not a map). Can I still use the MapUtils?

Reply:
I see, yes. In this case your solution is the way to go.